File: advent-of-code/src/main/kotlin/com/willmolloy/adventofcode/_2025/Day3.kt
package com.willmolloy.adventofcode._2025

import com.willmolloy.adventofcode.common.Day
import com.willmolloy.adventofcode.common.Input
import kotlin.math.max

/** https://adventofcode.com/2025/day/3 */
object Day3 : Day(2025, 3) {

  override fun part1(input: Input): Any{
    var res = 0

    for (line in input.lines()) {
      // try all the combos...
      var lineMax = 0
      for (i in 0 until line.length) {
        for (j in i + 1 until line.length) {
          val num = line[i].digitToInt() * 10 + line[j].digitToInt()
          lineMax = max(lineMax, num)
        }
      }

      res += lineMax
    }
    return res
  }

  override fun part2(input: Input): Any {
    // same but go to a depth of 12 rather than 2... thinking dfs
    // dfs too slow...
    // instead it can be greedy?? how?

    var res = 0L

    for (line in input.lines()) {
      var maxSum = 0L

      fun dfs(i: Int, turnedOn: Int, limit: Int, sum: Long){
        if (turnedOn == limit) {
          maxSum = max(maxSum, sum)
          return
        }

        for (j in i + 1 until line.length) {
          val nextSum = sum * 10 + line[j].digitToInt()
          dfs(j, turnedOn + 1, limit, nextSum)
        }
      }

      dfs(-1, 0, 12, 0)

      println("Max for: $line = $maxSum")
      res += maxSum
    }
    return res
  }
}

File: advent-of-code/src/main/scala/com/willmolloy/adventofcode/_2024/Day3.scala
package com.willmolloy.adventofcode._2024

import com.willmolloy.adventofcode.common.{Day, Input}

/**
 * https://adventofcode.com/2024/day/3
 */
object Day3 extends Day(2024, 3) {

  override def part1(input: Input): Int = {
    val p = """mul\((\d+),(\d+)\)""".r
    val matches = p.findAllMatchIn(input.readString())

    matches
      .map(m => {
        m.group(1).toInt * m.group(2).toInt
      })
      .sum
  }

  private def part1(line: String): Int = {
    val p = """mul\((\d+),(\d+)\)""".r
    val matches = p.findAllMatchIn(line)

    matches
      .map(m => {
        m.group(1).toInt * m.group(2).toInt
      })
      .sum
  }

  override def part2(input: Input): Int = {
    val p = """(?:do\(\)|^)(.*?)(?:don't\(\)|$)""".r
    val matches = p.findAllMatchIn(input.readString())

    matches
      .map(m => {
        part1(m.group(1))
      })
      .sum
  }
}
